src/Prefetch.cpp

```diff
@@ -129,39 +99,96 @@ class InjectPrefetch : public IRMutator2 {
     }
 
-    Stmt add_prefetch(string buf_name, const Parameter &param, const Box &box, Stmt body) {
-        // Construct the region to be prefetched.
-        Region bounds;
-        for (size_t i = 0; i < box.size(); i++) {
-            Expr extent = box[i].max - box[i].min + 1;
-            bounds.push_back(Range(box[i].min, extent));
+    Stmt visit(const Prefetch *op) override {
+        Stmt body = mutate(op->body);
+
+        const PrefetchDirective &p = op->prefetch;
+        Expr loop_var = Variable::make(Int(32), p.var);
+
+        // Add loop variable + prefetch offset to interval scope for box computation
+        Expr fetch_at = loop_var + p.offset;
+        map<string, Box> boxes_rw = boxes_touched(LetStmt::make(p.var, fetch_at, body));
+
+        // TODO(psuriana): Only prefetch the newly accessed data. We
+        // should subtract the box accessed during previous iteration
+        // from the one accessed during this iteration.
+
+        // TODO(psuriana): Add a new PrefetchBoundStrategy::ShiftInwards
+        // that shifts the base address of the prefetched box so that
+        // the box is completely within the bounds.
+        const auto &b = boxes_rw.find(p.name);
+        if (b != boxes_rw.end()) {
+            Box prefetch_box = b->second;
+            // Only prefetch the region that is in bounds.
+            Box bounds = get_buffer_bounds(b->first, b->second.size());
+            internal_assert(prefetch_box.size() == bounds.size());
+
+            if (p.strategy == PrefetchBoundStrategy::Clamp) {
+                prefetch_box = box_intersection(prefetch_box, bounds);
+            } else if (p.strategy == PrefetchBoundStrategy::GuardWithIf) {
+                Expr predicate = prefetch_box.used.defined() ? prefetch_box.used : const_true();
+                for (size_t i = 0; i < bounds.size(); ++i) {
+                    predicate = predicate && (prefetch_box[i].min >= bounds[i].min) &&
+                                (prefetch_box[i].max <= bounds[i].max);
+                }
+                prefetch_box.used = simplify(predicate);
+            } else {
+                internal_assert(p.strategy == PrefetchBoundStrategy::NonFaulting);
+                // Assume the prefetch won't fault when accessing region
+                // outside the bounds.
+            }
+
+            // Construct the region to be prefetched.
+            Region new_bounds;
+            for (size_t i = 0; i < prefetch_box.size(); i++) {
+                Expr extent = prefetch_box[i].max - prefetch_box[i].min + 1;
+                new_bounds.push_back(Range(simplify(prefetch_box[i].min), simplify(extent)));
+            }
+            Expr condition = op->condition;
+            if (prefetch_box.maybe_unused()) {
+                condition = simplify(prefetch_box.used && condition);
+            }
+            return Prefetch::make(op->name, op->types, new_bounds, op->prefetch, condition, std::move(body));
```

Comment:
Would it work to just generate an IfThenElse around the Prefetch, rather than embedding a condition in the Prefetch statement?

Reply:
No, since the Prefetch IR also has reference to 'body' (the actual producer-consumer which uses the prefetched region). The prefetch condition only applies for the prefetch call itself and not the 'body'.